include/lapack/lahqr_schur22.hpp

```diff
@@ -0,0 +1,222 @@
+                    if (scale >= safmx2)
+                    {
+                        sigma = sigma * safmn2;
+                        temp = temp * safmn2;
+                        continue;
+                    }
+                    if (scale <= safmn2)
+                    {
+                        sigma = sigma * safmx2;
+                        temp = temp * safmx2;
+                        continue;
+                    }
+                    break;
+                }
+                p = half * temp;
+                auto tau = lapy2(sigma, temp);
+                cs = sqrt(half * (one + abs(sigma) / tau));
+                sn = -(p / (tau * cs)) * copysign(one,sigma);
+                //
+                // Compute [aa bb] = [a b][cs -sn]
+                //         [cc dd] = [c d][sn  cs]
+                //
+                auto aa = a * cs + b * sn;
+                auto bb = -a * sn + b * cs;
+                auto cc = c * cs + d * sn;
+                auto dd = -c * sn + d * cs;
+                //
+                // Compute [a b] = [ cs sn][aa bb]
+                //         [c d] = [-sn cs][cc dd]
+                //
+                a = aa * cs + cc * sn;
+                b = bb * cs + dd * sn;
+                c = -aa * sn + cc * cs;
+                d = -bb * sn + dd * cs;
+
+                temp = half * (a + d);
+                a = temp;
+                d = temp;
+
+                if (c != zero)
+                {
+                    if (b != zero)
+                    {
+                        if (copysign(one,b) == copysign(one,c))
+                        {
+                            // Real eigenvalues: reduce to upper triangular form
+                            auto sab = sqrt(abs(b));
+                            auto sac = sqrt(abs(c));
+                            p = abs(c) * copysign(one,sab * sac);
+                            tau = one / sqrt(abs(b + c));
+                            a = temp + p;
+                            d = temp - p;
+                            b = b - c;
+                            c = zero;
+                            auto cs1 = sab * tau;
+                            auto sn1 = sac * tau;
+                            temp = cs * cs1 - sn * sn1;
+                            sn = cs * sn1 + sn * cs1;
+                            cs = temp;
+                        }
+                    }
+                }
+            }
+        }
+        // Store eigenvalues in s1 and s2
+        if (c != zero)
+        {
+            auto temp = sqrt(abs(b)) * sqrt(abs(c));
+            s1 = std::complex<T>(a, temp);
+            s2 = std::complex<T>(d, -temp);
+        }
+        else
+        {
+            s1 = a;
+            s2 = d;
+        }
+        return 0;
+    }
+
+    template <
+        typename T,
+        enable_if_t<is_complex<T>::value, bool> = true>
+    int lahqr_schur22(T &a, T &b, T &c, T &d, T &s1, T &s2, T &cs, T &sn)
+    {
+        return -1;
+    }
```

Comment:
I think we can remove this. If we keep this routine, `lahqr_schur22` will compile with complex types, and return `-1` at runtime. If we remove it, the error will be generated at compile time, which I think it is preferable.

Reply:
lahqr actually needs this. It catches the error to know that it shouldn't treat the 2x2 block as deflated and continue to apply shifts. I couldn't find a cleaner way other than writing lahqr twice.

Reply:
I think I saw the problem. My suggestion is:

```c++
diff --git a/include/lapack/lahqr.hpp b/include/lapack/lahqr.hpp
index e1d888f..4b4a9c2 100644
--- a/include/lapack/lahqr.hpp
+++ b/include/lapack/lahqr.hpp
@@ -214,7 +214,7 @@ namespace lapack
                     istart = ilo;
                     continue;
                 }
-                if (istart + 2 == istop)
+                if ( !is_complex<TA>::value && (istart + 2 == istop) )
                 {
                     // 2x2 block, normalize the block
                     TA cs;
diff --git a/include/lapack/lahqr_schur22.hpp b/include/lapack/lahqr_schur22.hpp
index b24b7bd..a65e1f3 100644
--- a/include/lapack/lahqr_schur22.hpp
+++ b/include/lapack/lahqr_schur22.hpp
@@ -50,9 +50,7 @@ namespace lapack
      *
      * @ingroup geev
      */
-    template <
-        typename T,
-        enable_if_t<!is_complex<T>::value, bool> = true>
+    template < typename T >
     int lahqr_schur22(T &a, T &b, T &c, T &d, std::complex<T> &s1, std::complex<T> &s2, T &cs, T &sn)
     {
 
@@ -74,6 +72,9 @@ namespace lapack
         const T safmn2 = pow(two, (int)(log(safmin / eps) / log(two)) / two);
         const T safmx2 = one / safmn2;
 
+        // Quick return
+        if( is_complex<T>::value ) return -1;
+
         if (c == zero)
         {
             // c is zero, the matrix is already in Schur form.
@@ -209,14 +210,6 @@ namespace lapack
         return 0;
     }
 
-    template <
-        typename T,
-        enable_if_t<is_complex<T>::value, bool> = true>
-    int lahqr_schur22(T &a, T &b, T &c, T &d, T &s1, T &s2, T &cs, T &sn)
-    {
-        return -1;
-    }
-
 } // lapack
 
 #endif // __LAHQR_SCHUR22_HH__
```

You can also use `lapack_error_if( is_complex<T>::value, -1 )` instead of `if( is_complex<T>::value ) return -1`, and this will change slightly the documentation of the method.

Reply:
I agree that checking for complex values in lahqr is better, but your suggestion doesn't compile on my pc, even if you return -1 if T is complex, the rest of that function would still need to compile for complex values. It's possible to do, but I think adding an overload is a lot cleaner.

Reply:
Hi. Ok, now I see the problem. The first routine won't work if T is complex. I am sorry for not noticing that. In this case, I agree there should be two versions: one for real types and other for complex types.

One comment: It is preferable to use `complex_type<T>` than `std::complex<T>` on the function signature. If nothing is said, `complex_type<T>` will be equivalent to `std::complex<T>`. There is the possibility, however, to use different structures for complex. The code is possibly not ready for that.